### DISCUSSION TITLE
Telegram bot

### DISCUSSION THREAD

**Discussion by ghost** [community/community#56984] [Programming Help] [answered]
### Body

Now I am developing ChatGPT telegram bot.
I am trying to get chat history of the group.

token = self.config['token']
        # get the chat ID
        chat_id = update.message.chat_id

        url = f'https://api.telegram.org/bot{token}/getUpdate'
        start_date = 1630540800
        end_date = 1684367829
        url += f'&oldest={start_date}&latest={end_date}'

        response = requests.get(url)
        data = response.json()
        print(data)
        output=""
        for temp in data['result']:
			message = temp['message']
			text = message['text']
			sender = message['from']['username']
			output+=(f'{sender}: {text}\n')

        
        # # get the chat history
		history = self.bot.get_history(chat_id)
        
        # # do something with the chat history
		for message in history:
			print(message.text)

        print(output)
        await update.message.reply_text(data)

But I can't get this history.
Please help me.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by SuperStar0718** [comment, marked as answer]
Hello
In the past, I've developed telegram bot integrated with ChatGPT using python.
At that moment, I got the chat history of the group using telegramclient class.

        api_id = 12345678
        api_hash = 'abcdefghijklmnopqrstuvwxyz'
        phone_number = '+123456798'
        entity = 'session'  # session_name
        client = TelegramClient(entity, api_id, api_hash)
        await client.start()
        await client.connect()
        # Log in to your account
        if not await client.is_user_authorized():
            await client.send_code_request(phone_number)
            await client.sign_in(phone_number, input('Enter the code you received: '))
        # chat_date = datetime.datetime.now()
        chat_date = datetime.datetime(year=2023, month=5, day=22)
        # Set up the date range
        start_date = datetime.datetime.now()- datetime.timedelta(hours=24)
        end_date = datetime.datetime.now()
        
        # Set up the offset IDs based on the date range
        start_message = await client.get_messages(chat_id, offset_date=start_date, limit=1)
        offset_id = start_message[0].id if start_message else 0
        end_message = await client.get_messages(chat_id, offset_date=end_date, limit=1)
        max_id = end_message[0].id if end_message else 0

        messages = await client.get_messages(chat_id, min_id=offset_id, max_id=max_id)

        users =await client.get_participants (chat_id)
        online_users=0
        for user in users:
            try:
                status = user.status.to_dict()
                print(f"{user.username}: {status['expires']}")
                online_users+=1
            except:
                pass

        await client.disconnect()
        output=""
        for message in reversed(messages):
            # print("message",message)
            output+=(f'"{message.text}" is sent by {message.sender.username}\n')

        # await update.message.reply_text(output)
        return output, online_users

This works very well.
I hope this will be helpful for you.

**Comment by inshanm21id** [comment]
Hello everyone, Please help me ....

I'd like to create a telegram bot that is used for a call center bot, and it's integrated with a spreadsheet to find the questions and answers of the FAQ for the customer.

I used the newest Python and I have already installed the telegram bot using PIP, but when I start the code, I get this problem:

Traceback (most recent call last):
  File "/Applications/bot_telegram/bot.py", line 6, in <module>
    from telegram.ext import Updater, CommandHandler, MessageHandler, Filters, CallbackContext
ImportError: cannot import name 'Filters' from 'telegram.ext' (/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/telegram/ext/__init__.py)

and here is the code that I use:

import logging
from datetime import datetime, time
import gspread
from oauth2client.service_account import ServiceAccountCredentials
from telegram import Update, ReplyKeyboardMarkup, ReplyKeyboardRemove
from telegram.ext import Updater, CommandHandler, MessageHandler, Filters, CallbackContext

# Configure logging
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)

# Initialize Google Sheets credentials
scopes = ['https://www.googleapis.com/auth/spreadsheets']
credentials = ServiceAccountCredentials.from_json_keyfile_name('credentials.json', scopes)

# Spreadsheet ID and FAQ worksheet
spreadsheet_id = '113ertLcHiP73WBH0GRWyq9B19hcAis5atEtqif4xFn0'
faq_sheet_name = 'FAQ Data'
chat_sheet_name = 'Chat History'

# Function to start the bot and greet the user
def start(update: Update, context: CallbackContext):
    user = update.effective_user
    context.bot.send_message(chat_id=update.effective_chat.id, text=f"Hello {user.first_name}! Welcome to our customer service. Please provide your name.")

# Function to handle user name submission
def handle_name(update: Update, context: CallbackContext):
    user = update.effective_user
    name = update.message.text

    # Save the user name to the context.user_data variable
    context.user_data['name'] = name

    context.bot.send_message(chat_id=update.effective_chat.id, text=f"Thank you, {name}! If you have any questions, please use the /faq command to see frequently asked questions.")

# Function to display the guide on how to use the bot
def show_guide(update: Update, context: CallbackContext):
    guide_text = "How to use the bot:\n\n" \
                 "/faq - View frequently asked questions\n" \
                 "Send another message - Ask another question\n" \
                 "/guide - Display the guide on how to use the bot"

    context.bot.send_message(chat_id=update.effective_chat.id, text=guide_text)

# Function to display FAQ questions and answers
def faq(update: Update, context: CallbackContext):
    # Open the Google Sheets spreadsheet
    gc = gspread.authorize(credentials)
    sheet = gc.open_by_key(spreadsheet_id).worksheet(faq_sheet_name)

    # Get all the question and answer data from the spreadsheet
    faq_data = sheet.get_all_records()

    # Create a message with the list of questions and answers
    faq_text = "Frequently Asked Questions:\n\n"
    for data in faq_data:
        faq_text += f"Q: {data['Question']}\n"
        faq_text += f"A: {data['Answer']}\n\n"

    context.bot.send_message(chat_id=update.effective_chat.id, text=faq_text)

# Function to handle user messages
def handle_message(update: Update, context: CallbackContext):
    user = update.effective_user
    message = update.message.text

    # Save the chat history to the Google Sheets spreadsheet
    gc = gspread.authorize(credentials)
    sheet = gc.open_by_key(spreadsheet_id).worksheet(chat_sheet_name)

    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    chat_data = [timestamp, user.id, message]
    sheet.append_row(chat_data)

    context.bot.send_message(chat_id=update.effective_chat.id, text="Thank you for your question! We will respond shortly.")

# Function to check the live chat time
def check_live_chat(context: CallbackContext):
    current_time = datetime.now().time()
    live_chat_start_time = time(8, 0)  # Live chat start time (8 AM)
    live_chat_end_time = time(17, 0)  # Live chat end time (5 PM)

    # If the current time is within the live chat time range
    if live_chat_start_time <= current_time <= live_chat_end_time:
        admin_chat_id = '@inshamid'  # Replace with the Telegram admin chat ID
        context.bot.send_message(chat_id=admin_chat_id, text="New question from a user.")

# Main function
def main():
    # Create Updater and Dispatcher objects
    updater = Updater(token='YOUR_TELEGRAM_BOT_TOKEN', use_context=True)
    dispatcher = updater.dispatcher

    # Add handler for the /start command
    start_handler = CommandHandler('start', start)
    dispatcher.add_handler(start_handler)

    # Add handler for user name submission
    name_handler = MessageHandler(Filters.text & ~Filters.command, handle_name)
    dispatcher.add_handler(name_handler)

    # Add handler for the /faq command
    faq_handler = CommandHandler('faq', faq)
    dispatcher.add_handler(faq_handler)

    # Add handler for other messages
    message_handler = MessageHandler(Filters.text & ~Filters.command, handle_message)
    dispatcher.add_handler(message_handler)

    # Add handler for the /guide command
    guide_handler = CommandHandler('guide', show_guide)
    dispatcher.add_handler(guide_handler)

    # Run the live chat every minute within the specified time range
    updater.job_queue.run_repeating(check_live_chat, interval=60, first=0)

    # Start the bot
    updater.start_polling()
    updater.idle()

if __name__ == '__main__':
    main()

**Reply by SuperStar0718** [reply]
So what do you want from me?
